src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.cs

```diff
@@ -43,15 +43,32 @@ public static partial class JsonSerializer
 
                     JsonTokenType tokenType = reader.TokenType;
 
+                    if (options.ReadReferenceHandling == ReferenceHandlingOnDeserialize.PreserveDuplicates && tokenType == JsonTokenType.PropertyName)
```

Comment:
This loop is extremely perf-sensitive. Please avoid comparing against the same tokenType more than once.